src/ssz/tree_view/list_composite.zig

```diff
@@ -0,0 +1,228 @@
+const hashing = @import("hashing");
+const Depth = hashing.Depth;
+const ListLengthUint = hashing.GindexUint;
+const Node = @import("persistent_merkle_tree").Node;
+const Gindex = @import("persistent_merkle_tree").Gindex;
+const isBasicType = @import("../type/type_kind.zig").isBasicType;
+
+const type_root = @import("../type/root.zig");
+const chunkDepth = type_root.chunkDepth;
+
+const tree_view_root = @import("root.zig");
+const BaseTreeView = tree_view_root.BaseTreeView;
+const CompositeChunks = @import("chunks.zig").CompositeChunks;
+
+/// A specialized tree view for SSZ list types with composite element types.
+/// Each element occupies its own subtree.
+pub fn ListCompositeTreeView(comptime ST: type) type {
+    comptime {
+        if (ST.kind != .list) {
+            @compileError("ListCompositeTreeView can only be used with List types");
+        }
+        if (!@hasDecl(ST, "Element") or isBasicType(ST.Element)) {
+            @compileError("ListCompositeTreeView can only be used with List of composite element types");
+        }
+    }
+
+    return struct {
+        base_view: BaseTreeView,
+        list_length_cache: ?ListLengthUint = null,
+
+        pub const SszType = ST;
+        pub const Element = ST.Element.TreeView;
+
+        const Self = @This();
+
+        const base_chunk_depth: Depth = @intCast(ST.chunk_depth);
+        const chunk_depth: Depth = chunkDepth(Depth, base_chunk_depth, ST);
+        const list_length_gindex: Gindex = Gindex.fromDepth(1, 1);
+        const Chunks = CompositeChunks(ST, chunk_depth);
+
+        const dirty_mask: ListLengthUint = @as(ListLengthUint, 1) << (@bitSizeOf(ListLengthUint) - 1);
+
+        comptime {
+            const max_len: usize = (@as(usize, 1) << (@bitSizeOf(ListLengthUint) - 1)) - 1;
+            if (ST.limit > max_len) {
+                @compileError("ListCompositeTreeView list length does not fit (limit too large for dirty-bit encoding)");
+            }
+        }
+
+        inline fn isDirtyLength(cached: ListLengthUint) bool {
+            return (cached & dirty_mask) != 0;
+        }
+
+        inline fn stripDirtyLength(cached: ListLengthUint) ListLengthUint {
+            return cached & ~dirty_mask;
+        }
+
+        inline fn tagDirtyLength(length_value: ListLengthUint) ListLengthUint {
+            std.debug.assert((length_value & dirty_mask) == 0);
+            return length_value | dirty_mask;
+        }
+
+        pub fn init(allocator: Allocator, pool: *Node.Pool, root: Node.Id) !Self {
+            return Self{
+                .base_view = try BaseTreeView.init(allocator, pool, root),
+                .list_length_cache = null,
+            };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.base_view.deinit();
+        }
+
+        pub fn commit(self: *Self) !void {
+            try self.base_view.commit();
+
+            if (self.list_length_cache) |cached| {
+                if (!isDirtyLength(cached)) return;
```

Comment:
when we only call `set()` for a list, we don't change the length
but the `self.base_view.data.root` only holds the root of the regular tree without `mix_in_length()`
we always have to bind to length node regardless
  - if dirty length, create new length node
  - if not, use the existing length node

this is same to what we have right now in typescript https://github.com/ChainSafe/ssz/blob/f5ed0b457333749b5c3f49fa5eafa096a725f033/packages/ssz/src/type/arrayBasic.ts#L51

also looks like there is no unit test for this case (with set() only)

Reply:
https://github.com/ChainSafe/lodestar-z/blob/main/src/ssz/type/list.zig#L278 I thought it already include `length node`?